### PR TITLE
Cit checkout flow improvements

### PR DESCRIPTION
**Pull request by michelleinez**

### CODE REVIEW

File: src/components/Checkout/BasketItem.vue
Comment (mcstover) [inline src/components/Checkout/BasketItem.vue:190]: I think this can be replaced with `tw-cursor-pointer`...

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (mcstover) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:134]: I think there will be unintended consequences of allowing this into the DOM all the time. It will likely start to initialize the checkout before we have all the information for that process. It will also load up more async code on page load instead of on demand.

File: src/components/Checkout/OrderTotals.vue
Comment (mcstover) [inline src/components/Checkout/OrderTotals.vue:25]: DonationItem only emits the `refreshtotals` event without the `-` so this will need to be reverted.

File: src/components/Checkout/BasketItem.vue
Comment (dyersituations) [inline src/components/Checkout/BasketItem.vue:87]: This can just be a `v-else`

File: src/components/CorporateCampaign/CampaignProgressBar.vue
Comment (dyersituations) [inline src/components/CorporateCampaign/CampaignProgressBar.vue:75]: I think this is a duplicate component, but it looks like the SCSS could be replaced with TW.

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (dyersituations) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1324]: Do we want to use `logFormatter` here and add an error log?

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (dyersituations) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1319]: Is this success log needed?

**Comment by dyersituations** [comment]
@michelleinez Let us know if there's a way we can test this flow locally. My test user isn't part of a company with credit, so I don't see this flow locally.

File: package.json
Comment (michelleinez) [reply package.json:99]: Yeah, I noticed this after pushing the branch up...  This is a combination of commits from myself and others and I will need to revert this but was up too late and just wanted to get feedback on the rest of the PR asap so I could work on the rest of it too, so I put it up anyway with the intention to update it later.

File: src/components/Checkout/OrderTotals.vue
Comment (mcstover) [inline src/components/Checkout/OrderTotals.vue:85]: Folks can add a Redemption Code or Have lending rewards or even UPC codes outside of managed lending campaigns and they are not required to use that credit for their current loan purchase. This means there can still be an available amount that isn't applied.

Thus, we'll need to keep some display of the available credit if that computed property has a value.

File: src/components/Checkout/BasketItem.vue
Comment (michelleinez) [reply src/components/Checkout/BasketItem.vue]: done

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: done

File: src/components/Checkout/OrderTotals.vue
Comment (michelleinez) [reply src/components/Checkout/OrderTotals.vue]: done

File: package.json
Comment (michelleinez) [reply package.json]: done

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: not needed, removed it

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: done

File: src/components/Checkout/BasketItem.vue
Comment (michelleinez) [reply src/components/Checkout/BasketItem.vue]: done

File: src/components/CorporateCampaign/CampaignProgressBar.vue
Comment (michelleinez) [reply src/components/CorporateCampaign/CampaignProgressBar.vue]: done

**Comment by michelleinez** [comment]
> @michelleinez Let us know if there's a way we can test this flow locally. My test user isn't part of a company with credit, so I don't see this flow locally.

@dyersituations you can use https://localhost:8888/cc/zuora?upc=zuoracustomers

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (mcstover) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:745]: We may want to rename this to `handleScrollToLoans` to prevent confusion

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (mcstover) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1225]: This `return` just means "continue" in this context...so prob not needed.

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (mcstover) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1345]: Could we please namespace this method "updateBasketItem", "completeRemoveBasketItem" and "trackItemRemoved" as they seem to be tightly tied to the new remaining credit allocation mechanisms.

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (mcstover) [inline src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1357]: I'm a bit concerned that this is a duplicate behavior. The actual removal of an item from basket and it's tracking will have been handled by the BasketItem component. Does this get called from some context outside of a BasketItem?

**Comment by dyersituations** [comment]
These errors might be unrelated to this work, but I'm seeing the first error on page load, the second error whenever add to basket is clicked, and the third if the user isn't logged in when adding.

image

**Comment by dyersituations** [comment]
Maybe this won't be an issue on prod, but clicking the basket link in the header while the promotion is validating (which takes a number of seconds when I load the page locally), adds `#show-basket` to the browser URL but doesn't show the basket. Then the basket link doesn't do anything after the promotion is validated, since the URL doesn't get changed again, it already has `#show-basket`.

**Comment by dyersituations** [comment]
It looks like if there's no promo credit, the checkout button updates the URL but does nothing. Should it be removed when there's no credit?

image

**Comment by dyersituations** [comment]
> @dyersituations you can use https://localhost:8888/cc/zuora?upc=zuoracustomers

Thanks! I went through the whole process, and now the ups doesn't work again. Is it a one-use? I don't need another code at the moment, but I thought I should mention that it shows as used for me now.

**Comment by mcstover** [comment]
Regarding the errors @dyersituations mentioned. The auth related log is common if you have not yet logged in. However, the Invariant error is a problem we need to resolve if it's present.

As for UPC code usage, they can be configured to only allow 1 UPC usage per user account so I suspect that is why you can't reapply that code. Or, it may be that this test account has run out of funds.

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1357]: @mcstover yes, it does.  When the basket is updated or when it is loaded, the CCLandingPage compares the amount of credit that is still available to the amount applied.  If there is leftover credit available to allocate towards a loan, the code chooses a loan from the carousel and applies that credit to it.  It also remembers in a cookie which loan had remaining credit allocated to it so that if a user adds more credit to a different loan, the credit is removed from that original loan that had credit allocated to it.  That removal process happens with no intervention from the user and occurs when the basket state is updated, so it wouldn't be something that originates from the "RemoveBasketItem" component in the BasketItem...

Comment (mcstover) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue:1357]: Ok thank you, that makes sense. Just wanted to make sure we weren't calling the mutation or tracking the removal twice.

File: src/components/Checkout/OrderTotals.vue
Comment (mcstover) [inline src/components/Checkout/OrderTotals.vue:450]: Do we need to revert these changes? What does this copy look like now on desktop?

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: done

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: done

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: done

File: src/components/Checkout/OrderTotals.vue
Comment (michelleinez) [reply src/components/Checkout/OrderTotals.vue]: Okay got it -- next thing I plan to work on is incorporating a flag that enables the order totals component to know if it's on a campaign page so we can retain the way these totals are displayed but swap it out for the new design on the cc pages.  But for now I've reinstated the old order totals so that kiva.org checkout won't be affected.

File: src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue
Comment (michelleinez) [reply src/pages/LandingPages/CorporateCampaign/CCLandingPage.vue]: removed them

File: src/components/Checkout/OrderTotals.vue
Comment (michelleinez) [reply src/components/Checkout/OrderTotals.vue:450]: reverted, done

File: src/components/Checkout/OrderTotals.vue
Comment (michelleinez) [reply src/components/Checkout/OrderTotals.vue:85]: okay, I've since added that flag now.

**Comment by michelleinez** [comment]
> These errors might be unrelated to this work, but I'm seeing the first error on page load, the second error whenever add to basket is clicked, and the third if the user isn't logged in when adding.
> 
> image

Added a line in the basketUtils.js file that suppresses this error by providing a temp-id.

File: src/components/Checkout/TeamAttribution.vue
Comment (mcstover) [inline src/components/Checkout/TeamAttribution.vue:104]: From what I can see, this won't do anything. The TeamAttribution.vue component only lives in the BasketItem.vue file and the component usage doesn't handle any emitted events at this time. You would need to add the `@refreshtotals="doSomething"` to `<team-attribution>`.

File: src/components/CorporateCampaign/CampaignLoanRow.vue
Comment (mcstover) [inline src/components/CorporateCampaign/CampaignLoanRow.vue:40]: How are you handling showing checkout after the add to basket event now?

Comment (mcstover) [reply src/components/CorporateCampaign/CampaignLoanRow.vue:40]: These are the modifiers that set the hash to open the basket and I think they ensure the basket is "re-opened" when returning from login as well...

Comment (michelleinez) [reply src/components/CorporateCampaign/CampaignLoanRow.vue:40]: We show it in the campaign progress bar instead and deactivate the loan card buttons.

File: src/components/Checkout/TeamAttribution.vue
Comment (michelleinez) [reply src/components/Checkout/TeamAttribution.vue:104]: I guess I was just trying to be thorough and make sure every instance of event emission related to refresh totals is without the dash.  Do I need to revert or can I just let this be?

Comment (michelleinez) [reply src/components/Checkout/TeamAttribution.vue]: done

**Comment by dyersituations** [comment]
I was on the cc page then went directly to checkout. This link does nothing in the checkout page:
image

**Comment by dyersituations** [comment]
It looks like the basket isn't checked if loans fulfill the promo until the checkout dialog is shown.

1) open page with $50 credit: http://localhost:8888/cc/zuora?upc=zuoracustomers
2) add 1 loans
3) click "checkout now" button -> notice basket now has 2 loans
4) add second loan -> notice basket now has 3 loans
5) go directly to checkout -> notice there are 3 loans and $25 due
6) return to cc/zuora
7) click "checkout now" button -> notice basket is now 2 loans instead of 3

It looks like the basket status needs to be updated when loans are added to the basket, instead of when "checkout now" is clicked.

**Comment by michelleinez** [comment]
> Padding missing around login button:
> 
> image

Thanks for finding these bugs.  Regarding the login padding, I do have plans to fix that in a separate commit but I don't think that fix belongs here.  It's not a regression, it's always been like that since I joined the team.

**Comment by dyersituations** [comment]
> Thanks for finding these bugs. Regarding the login padding, I do have plans to fix that in a separate commit but I don't think that fix belongs here. It's not a regression, it's always been like that since I joined the team.

No problem! Yeah I'm not sure which bugs are feature creep, just let us know. I'm also cool if follow-up tickets are created in general, if that works with your team's workflow, and can approve based on follow-up work.

**Comment by michelleinez** [comment]
> It looks like the basket isn't checked if loans fulfill the promo until the checkout dialog is shown.
> 
> 1. open page with $50 credit: http://localhost:8888/cc/zuora?upc=zuoracustomers
> 2. add 1 loans
> 3. click "checkout now" button -> notice basket now has 2 loans
> 4. add second loan -> notice basket now has 3 loans
> 5. go directly to checkout -> notice there are 3 loans and $25 due
> 6. return to cc/zuora
> 7. click "checkout now" button -> notice basket is now 2 loans instead of 3
> 
> It looks like the basket status needs to be updated when loans are added to the basket, instead of when "checkout now" is clicked.
As for these other two, I'll try to repro them and address them.

**Comment by michelleinez** [comment]
> I was on the cc page then went directly to checkout. This link does nothing in the checkout page: image

@dyersituations Did you clear cache when you ran this?  I know this was an issue before, but I'm not able to make the link fail now.  Having trouble reproducing.

**Comment by dyersituations** [comment]
> @dyersituations Did you clear cache when you ran this? I know this was an issue before, but I'm not able to make the link fail now. Having trouble reproducing.

Yeah I had restarted my local dev server. I can test again, but it looks like the promo code has ended.

**Comment by michelleinez** [comment]
> > @dyersituations Did you clear cache when you ran this? I know this was an issue before, but I'm not able to make the link fail now. Having trouble reproducing.
> 
> Yeah I had restarted my local dev server. I can test again, but it looks like the promo code has ended.

I just reinstated it.

**Comment by dyersituations** [comment]
I just pulled the branch, restarted my local dev server to clear the cache, and I'm still seeing the checkout page link not working.

1) Open page w/ applied promo
2) Add one loan
3) Go directly to /checkout
4) link in this screenshot doesn't do anything

image

**Comment by dyersituations** [comment]
Not sure how common the basket error is on the cc pages, but platform frontend recently added some fixes for this case that could be ported to this page. Definitely a follow-up ticket, but I thought it was worth mentioning, since I just ran into the error locally.
image

**Comment by dyersituations** [comment]
@michelleinez Could we get the promo enabled again?

**Comment by michelleinez** [comment]
> @michelleinez Could we get the promo enabled again?
Hey Casey, thanks for being so on top of the reviews here, I will re-request another review in git when I'm ready for you to take a look again, but for now I'm not ready.  I'm having trouble this time getting the promo to show up and not sure what is going on, still troubleshooting.

**Comment by michelleinez** [comment]
> > @michelleinez Could we get the promo enabled again?
> > Hey Casey, thanks for being so on top of the reviews here, I will re-request another review in git when I'm ready for you to take a look again, but for now I'm not ready.  I'm having trouble this time getting the promo to show up and not sure what is going on, still troubleshooting.

@dyersituations Okay I think I'm ready now.  You can test with http://localhost:8888/cc/deloitte?upc=deloitte

**Comment by michelleinez** [comment]
> > > @michelleinez Could we get the promo enabled again?
> > > Hey Casey, thanks for being so on top of the reviews here, I will re-request another review in git when I'm ready for you to take a look again, but for now I'm not ready.  I'm having trouble this time getting the promo to show up and not sure what is going on, still troubleshooting.
> 
> @dyersituations Okay I think I'm ready now. You can test with http://localhost:8888/cc/deloitte?upc=deloitte

@dyersituations Oh also, before testing, clear site data first in your browser.  That's usually what causes the basket issue.  But if there are fixes for it that can be ported over as well, that's good to know.  I can make a ticket for us to get that to happen.

**Comment by dyersituations** [comment]
If you click on the "..." in the carousel, it doesn't highlight the selected step, where it does when navigating with the arrows

image

**Comment by dyersituations** [comment]
We fixed this for the new loan cards, but these cards have varying heights in the carousel:

image

**Comment by dyersituations** [comment]
Pager adds to query string, but loading with query string doesn't apply to grid paging

image

**Comment by dyersituations** [comment]
If I select 2 loans, a third is shown when clicking "Checkout", but the header only mentions 2. If I then go to the checkout page directly, that value changes to 3. Refreshing the campaign page also changes the basket number to 3.

**Comment by dyersituations** [comment]
@michelleinez Most of the screenshots I just added are minor, the only one I think might be more important is the basket number differences. Is the current behavior described in one of the comments expected?

File: src/components/Checkout/InContext/InContextCheckout.vue
Comment (dyersituations) [inline src/components/Checkout/InContext/InContextCheckout.vue:32]: Doesn't need to change, but this can be an object:
```
class="in-context-login"
:class={ 'tw-text-right':  !isCorporateCampaignPage }
```

File: src/components/Checkout/InContext/InContextCheckout.vue
Comment (dyersituations) [inline src/components/Checkout/InContext/InContextCheckout.vue:180]: This could be extracted into a shared util

File: src/components/LoanCards/KivaClassicBasicLoanCardExp.vue
Comment (dyersituations) [inline src/components/LoanCards/KivaClassicBasicLoanCardExp.vue:209]: This prop can probably be removed, since this card isn't being used, and this prop isn't being used.

**Comment by michelleinez** [comment]
> @michelleinez Most of the screenshots I just added are minor, the only one I think might be more important is the basket number differences. Is the current behavior described in one of the comments expected?

When was the last time you tested this?
I'm not experiencing this issue anymore for the version with the latest changes.

**Comment by michelleinez** [comment]
> > @michelleinez Most of the screenshots I just added are minor, the only one I think might be more important is the basket number differences. Is the current behavior described in one of the comments expected?
> 
> When was the last time you tested this? I'm not experiencing this issue anymore for the version with the latest changes.

Oh but to answer your question, no the basket count should reflect the number of items in the basket.

**Comment by dyersituations** [comment]
> When was the last time you tested this? I'm not experiencing this issue anymore for the version with the latest changes.

I just checked again locally in an incognito browser, and it worked like expected!

**Review by dyersituations — APPROVED** [review]
Approved pending merge conflict resolution.

**Comment by michelleinez** [comment]
> Approved pending merge conflict resolution.

Thanks, I'll have to wait until after April 1st to merge due to the ongoing VISA campaign, so I'm thinking to wait until end of march to resolve these so I can address all of the merge conflicts in one commit.

**Pull request merged**
